addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,304 @@
+    }
+}
 
 /**
- * @private
- * @param {string} modelName
- * @returns {Object}
+ * Adds the provided hooks to the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to add the hooks.
+ * @param {Object} hooks Hooks to be added.
  */
-function _getEntryFromModelName(modelName) {
-    if (!registry[modelName]) {
-        registry[modelName] = {
-            dependencies: [],
-            factory: undefined,
-            name: modelName,
-            patches: [],
-        };
-    }
-    return registry[modelName];
+export function addLifecycleHooks(modelName, hooks) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot add lifecycle hooks to model "${modelName}": model must be registered before lifecycle hooks can be added.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [name, handler] of Object.entries(hooks)) {
+        if (typeof handler !== 'function') {
+            throw new Error(`Cannot add lifecycle hook "${name}" to model "${modelName}": "${name}" must be a function.`);
+        }
+        if (definition.get('lifecycleHooks').has(name)) {
+            throw new Error(`Cannot add lifecycle hook "${name}" to model "${modelName}": hook already exists.`);
+        }
+        definition.get('lifecycleHooks').set(name, handler);
+    }
 }
 
 /**
- * @private
- * @param {string} modelName
- * @param {string} patchName
- * @param {Object} patch
- * @param {Object} [param3={}]
- * @param {string} [param3.type='instance'] 'instance', 'class', 'field' or 'identifyingFields'
+ * Adds the provided model getters to the model specified by the `modelName`.
+ *
+ * @deprecated
+ * @param {string} modelName The name of the model to which to add the model getters.
+ * @param {Object} modelGetters Model getters to be added.
  */
-function _registerPatchModel(modelName, patchName, patch, { type = 'instance' } = {}) {
-    const entry = _getEntryFromModelName(modelName);
-    Object.assign(entry, {
-        patches: (entry.patches || []).concat([{
-            name: patchName,
-            patch,
-            type,
-        }]),
-    });
+ export function addModelGetters(modelName, modelGetters) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot add record getters to model "${modelName}": model must be registered before record getters can be added.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [getterName, getter] of Object.entries(modelGetters)) {
+        if (typeof getter !== 'function') {
+            throw new Error(`Cannot add model getter "${getterName}" to model "${modelName}": "${getterName}" must be a function.`);
+        }
+        if (definition.get('modelGetters').has(getterName)) {
+            throw new Error(`Cannot add model getter "${getterName}" to model "${modelName}": getter already exists.`);
+        }
+        definition.get('modelGetters').set(getterName, getter);
```

Comment:
Ideally need to add check for name conflict between model getter/method and record getter/method/fields (both ways)